### ISSUE TITLE
Unable to set optimization function and constraint(s) with the same function

### ISSUE THREAD

**Issue by damianoazzolini** [stevengj/nlopt#372]
Hi, this is my issue. Consider the following program where i have 2 variables, call them `a` and `b` (both between 0 and 1). The objective function i want to optimize (minimize) is `a+b`. The only inequality constraint is `0.3 -a*b < 0` (i.e., `a*b > 0.3`). I use the same function in both `nlopt_set_min_objective` and `nlopt_add_inequality_constraint,` and i use a flag to discriminate between the two cases (minimization ad constraint functions). Here is the C code.
```C
double constr(unsigned n, const double *x, double *grad, void *data) {
    int *val = (int *)data;
    if(grad) {
        if(val == 0) {
            grad[0] = 1;
            grad[1] = 1;
        }
        else {
            grad[0] = -x[1];
            grad[1] = -x[0];
        }

    }

    if(val == 0) {
        return x[0] + x[1];
    }
    else {
        return 0.3 - (x[0] * x[1]);
    }
}
```
 I call them with like:
```C
int sel = 0;
nlopt_set_min_objective(opt, constr, &sel);
sel = 1;
nlopt_add_inequality_constraint(opt, constr, &sel, 1e-8);
```
so, for this example, variable `sel` is used to discriminate between the two cases. However, this not work as expected, since i get `found minimum at f(1,1) = -0.7` while i would expect `found minimum at f(0.547723,0.547723) = 1.095445123`. I've obtained this correct result by creating two different functions, one for the constraint and one for the objective. In both cases, i've selected `NLOPT_LD_MMA` as algorithm, so gradients are used.

To sum up: can i use the same function for both objective and constraint(s)? Furthermore, can i use the same function for more than one constraint (suppose i need to add also another constraint that involves the product of the two variables)? 
This because, in my problem, i know the number of constraints only at *RUNTIME*, so i cannot create n different functions at runtime. 
I would like to reuse the same function but with different parameters. Is it possible?

Thanks, sorry for the long issue.
